hathor/consensus/poa/poa_block_producer.py

```diff
@@ -0,0 +1,172 @@
+        self._poa_signer = poa_signer
+        self._signer_index = self._calculate_signer_index(self._poa_settings, self._poa_signer)
+        self._last_seen_best_block: PoaBlock | None = None
+
+        self._started_producing = False
+        self._start_producing_lc = LoopingCall(self._start_producing)
+        self._start_producing_lc.clock = self._reactor
+
+        self._schedule_block_lc = LoopingCall(self._schedule_block)
+        self._schedule_block_lc.clock = self._reactor
+
+    def start(self) -> None:
+        self._start_producing_lc.start(_WAIT_SYNC_DELAY)
+        self._schedule_block_lc.start(self._settings.AVG_TIME_BETWEEN_BLOCKS)
+
+    def stop(self) -> None:
+        if self._start_producing_lc.running:
+            self._start_producing_lc.stop()
+
+        if self._schedule_block_lc.running:
+            self._schedule_block_lc.stop()
+
+    @staticmethod
+    def _calculate_signer_index(settings: PoaSettings, poa_signer: PoaSigner) -> int:
+        """Return the signer index for the given private key."""
+        public_key = poa_signer.get_public_key()
+        public_key_bytes = get_public_key_bytes_compressed(public_key)
+        sorted_signers = sorted(settings.signers)
+        try:
+            return sorted_signers.index(public_key_bytes)
+        except ValueError:
+            raise ValueError(f'Public key "{public_key_bytes.hex()}" not in list of PoA signers')
+
+    def _start_producing(self) -> None:
+        """Start producing new blocks."""
+        if not self._manager.can_start_mining():
+            # We're syncing, so we'll try again later
+            self._log.warn('cannot start producing new blocks, node not synced')
+            return
+
+        self._log.info('started producing new blocks')
+        self._started_producing = True
+        self._start_producing_lc.stop()
+
+    def _schedule_block(self) -> None:
+        """Schedule propagation of a new block."""
+        previous_block = self._manager.tx_storage.get_best_block()
+        assert isinstance(previous_block, PoaBlock)
+        if not self._started_producing or previous_block == self._last_seen_best_block:
+            return
+
+        self._last_seen_best_block = previous_block
+        now = self._reactor.seconds()
+        expected_timestamp = self._expected_block_timestamp(previous_block)
+        propagation_delay = 0 if expected_timestamp < now else expected_timestamp - now
+
+        self._reactor.callLater(propagation_delay, self._produce_block, previous_block)
+        self._log.debug(
+            'scheduling block production',
+            previous_block=previous_block.hash_hex,
+            previous_block_height=previous_block.get_height(),
+            delay=propagation_delay,
+        )
+
+    def _produce_block(self, previous_block: PoaBlock) -> None:
+        """Create and propagate a new block."""
+        block_templates = self._manager.get_block_templates(parent_block_hash=previous_block.hash)
```

Comment:
@msbrogli commented on https://github.com/HathorNetwork/hathor-core-side-dag/pull/11#discussion_r1634979033:

> Why don't you get the previous block from `self._manager.tx_storage.get_best_block()`? Why are you getting this information when a block is scheduled (instead of when it is generated)?

Reply:
But won't peers reject out-of-turn blocks in most cases? I mean, can a misconfigured peer generate and propagate out-of-turn blocks?

Reply:
What if `previous_block != self._manager.tx_storage.get_best_block()`? Should we keep producing our block? Or should we wait for our next turn?

Reply:
Wouldn't it be better to accept out-of-turn blocks, even if the recovery strategy is inefficient it should at least be correct and we can implement a better one in the future, very possibly in a backwards compatible way. At least accepting out-of-turn blocks would make it easier to make a backwards compatible improvement.

Reply:
@msbrogli 

> But won't peers reject out-of-turn blocks in most cases?

What do you mean by "reject"? The blocks will be accepted, but voided if we receive an in-turn block for the same height.

> I mean, can a misconfigured peer generate and propagate out-of-turn blocks?

A well-configured peer will generate and propagate out-of-turn blocks when the in-turn signer is not online. A misconfigured peer will only do something different in a modified full node.

> What if previous_block != self._manager.tx_storage.get_best_block()? Should we keep producing our block? Or should we wait for our next turn?

As explained in the previous comment, in that case it'll simply be voided and not propagated.

Reply:
@jansegre 

> Wouldn't it be better to accept out-of-turn blocks?

We _do_ accept out-of-turn blocks

Reply:
Thanks!